File: pipeline/build_rtl.go
package main

// build-rtl — multi-language RTL demo emission, ported from
// tools/build-rtl.mjs. Reads src/registry/rtl-translations.json, substitutes
// each ar value with the target language's, patches lang/dir attributes,
// and injects the theme pre-paint script (the same one src/docs/theme-prepaint.mjs
// ships — ported into prepaint.go for the Go tree).
//
// Output: docs/demos/<name>-{he,en,fa}.html + dist/components/<name>-… + the
// language manifest build/rtl-langs.json. Nothing is written on any failure.

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

// substituteAndPatch mirrors tools/rtl-lib.mjs. Order of the two phases is
// load-bearing: ar→X substitution first (longest-value ordering avoids a
// "Loading" / "Loading…" prefix collision), then the lang/dir attribute
// patch.
func substituteAndPatch(arabicHTML string, translations map[string]struct {
	Dir    string            `json:"dir"`
	Values map[string]string `json:"values"`
}, fromLang, toLang string, toValues map[string]string, toDirOverride string) string {
	type langEnt struct {
		Dir    string
		Values map[string]string
	}
	// typed view
	tr := map[string]langEnt{}
	for k, v := range translations {
		tr[k] = langEnt{Dir: v.Dir, Values: v.Values}
	}
	from := tr[fromLang].Values
	// keys sorted by descending from-value length so longer strings match
	// first (longest-prefix problem); missing from-side sorts last
	keys := make([]string, 0, len(toValues))
	for k := range toValues {
		keys = append(keys, k)
	}
	sort.Slice(keys, func(i, j int) bool {
		return len(from[keys[j]]) < len(from[keys[i]])
	})
	var unmatched []string
	out := arabicHTML
	for _, key := range keys {
		fromVal := from[key]
		toVal := toValues[key]
		if fromVal != "" && toVal != "" && fromVal != toVal {
			if !strings.Contains(out, fromVal) {
				unmatched = append(unmatched, key)
				continue
			}
			out = strings.ReplaceAll(out, fromVal, toVal)
		} else if toVal != "" && fromVal == "" {
			unmatched = append(unmatched, key+"(no "+fromLang+" source)")
		}
	}
	if len(unmatched) > 0 {
		fmt.Fprintf(os.Stderr, "warn %s: translation keys not found in HTML (left as %s): %s\n",
			toLang, fromLang, strings.Join(unmatched, ", "))
	}
	langDir := toDirOverride
	if langDir == "" {
		langDir = tr[toLang].Dir
	}
	if langDir == "" {
		langDir = "ltr"
	}
	// <html lang>: replace or inject
	if regexp.MustCompile(`<html[^>]*\slang="`).MatchString(out) {
		out = regexp.MustCompile(`(<html[^>]*\slang=")[^"]*(")`).
			ReplaceAllString(out, `${1}`+toLang+`${2}`)
	} else {
		out = regexp.MustCompile(`<html(\s[^>]*)?>`).
			ReplaceAllStringFunc(out, func(m string) string {
				if !strings.HasSuffix(m, ">") {
					return m
				}
				return m[:len(m)-1] + ` lang="` + toLang + `">`
			})
	}
	// every dir attribute (attribute-boundary anchored: a bare global would
	// also rewrite data-dir="ltr")
	out = regexp.MustCompile(`([\s"'])dir="(rtl|ltr)"`).
		ReplaceAllString(out, `${1}dir="`+langDir+`"`)
	return out
}


func runBuildRtl() int {
	dictB, err := os.ReadFile("src/registry/rtl-translations.json")
	if err != nil {
		fmt.Fprintln(os.Stderr, "build-rtl:", err)
		return 1
	}
	var dict map[string]map[string]struct {
		Dir    string            `json:"dir"`
		Values map[string]string `json:"values"`
	}
	if err := json.Unmarshal(dictB, &dict); err != nil {
		fmt.Fprintln(os.Stderr, "build-rtl: dict:", err)
		return 1
	}
	tiersB, err := os.ReadFile("src/registry/tiers.json")
	if err != nil {
		fmt.Fprintln(os.Stderr, "build-rtl:", err)
		return 1
	}
	var tiers map[string]struct {
		Tier string `json:"tier"`
		Emit bool   `json:"emit"`
	}
	json.Unmarshal(tiersB, &tiers)
	shipped := func(name string) bool {
		t, ok := tiers[strings.TrimSuffix(name, "-rtl")]
		return ok && (t.Emit || t.Tier == "static" || t.Tier == "kernel" || t.Tier == "trivial-js")
	}

	for _, d := range []string{"docs/demos", "dist/components", "build"} {
		if err := os.MkdirAll(d, 0o755); err != nil {
			fmt.Fprintln(os.Stderr, "build-rtl:", err)
			return 1
		}
	}

	persian := map[string]string{
		"paymentTitle":       "پرداخت موفق",
		"paymentDescription": "پرداخت حضرت به مبلغ ۲۹.۹۹ دلار با موفقیت انجام شد. رسید نیز به نشانی پست الکترونیکی شما ارسال گردید.",
		"featureTitle":       "ویژگی جدید موجود است",
		"featureDescription": "ما پشتیبانی از حالت تیره را به سیستم افزوده‌ایم. می‌توانید این قابلیت را از بخش تنظیمات حساب کاربری خود فعال نمایید.",
	}

	var names []string
	for n := range dict {
		names = append(names, n)
	}
	sort.Strings(names)

	type pending struct{ path, html string }
	var pendingWrites []pending
	var failures, skipped []string
	manifest := map[string][]string{}
	emitted := 0

	for _, name := range names {
		existing := "docs/demos/" + name + ".html"
		b, err := os.ReadFile(existing)
		if err != nil {
			if shipped(name) {
				failures = append(failures, name+": no "+existing+" to translate from, but the component ships")
			} else {
				skipped = append(skipped, name)
			}
			continue
		}
		ar := dict[name]
		translations := map[string]struct {
			Dir    string            `json:"dir"`
			Values map[string]string `json:"values"`
		}{}
		for k, v := range ar {
			translations[k] = v
		}

		langs := []string{"ar"}
		for _, lang := range []string{"he", "en"} {
			entry, ok := ar[lang]
			if !ok {
				continue
			}
			html := injectPrePaint(substituteAndPatch(string(b), translations, "ar", lang, entry.Values, ""))
			for _, dst := range []string{
				filepath.Join("docs/demos", name+"-"+lang+".html"),
				filepath.Join("dist/components", name+"-"+lang+".html"),
			} {
				pendingWrites = append(pendingWrites, pending{dst, html})
			}
			emitted++
			langs = append(langs, lang)
		}
		if name == "alert-rtl" {
			html := injectPrePaint(substituteAndPatch(string(b), translations, "ar", "fa", persian, "rtl"))
			for _, dst := range []string{
				filepath.Join("docs/demos", name+"-fa.html"),
				filepath.Join("dist/components", name+"-fa.html"),
			} {
				pendingWrites = append(pendingWrites, pending{dst, html})
			}
			emitted++
			langs = append(langs, "fa")
		}
		manifest[name] = langs
	}

	if len(failures) > 0 {
		for _, f := range failures {
			i := strings.Index(f, ": ")
			fmt.Fprintf(os.Stderr, "FAIL [%s]: %s\n", f[:i], f[i+2:])
		}
		fmt.Fprintf(os.Stderr, "FAIL  build-rtl (%d previews could not be built) — nothing written\n", len(failures))
		return 1
	}
	for _, p := range pendingWrites {
		if err := os.WriteFile(p.path, []byte(p.html), 0o644); err != nil {
			fmt.Fprintln(os.Stderr, "build-rtl:", err)
			return 1
		}
	}
	mb, err := json.MarshalIndent(manifest, "", "  ")
	if err != nil {
		fmt.Fprintln(os.Stderr, "build-rtl:", err)
		return 1
	}
	if err := os.WriteFile("build/rtl-langs.json", mb, 0o644); err != nil {
		fmt.Fprintln(os.Stderr, "build-rtl:", err)
		return 1
	}
	fmt.Printf("build-rtl: %d language variants emitted (excluding ar default) + manifest for %d previews", emitted, len(manifest))
	if len(skipped) > 0 {
		fmt.Printf(" (%d dictionaries have no shipped page: %s)", len(skipped), strings.Join(skipped, ", "))
	}
	fmt.Println()
	return 0
}

File: pipeline/src/emit/build_rtl.rs
//! Port of pipeline/build_rtl.go — multi-language RTL demo emission. Reads
//! src/registry/rtl-translations.json, substitutes each ar value with the
//! target language's, patches lang/dir attributes, and injects the theme
//! pre-paint script. Nothing is written on any failure.

use super::prepaint::inject_pre_paint;
use regex::Regex;
use serde::Deserialize;
use std::collections::{BTreeMap, HashMap};
use std::sync::{LazyLock, OnceLock};

static RE_HTML_HAS_LANG: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r#"<html[^>]*[\t\n\f\r ]lang=""#).unwrap());
static RE_HTML_LANG_ATTR: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r#"(<html[^>]*[\t\n\f\r ]lang=")[^"]*(")"#).unwrap());
static RE_HTML_OPEN_TAG: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"<html(\s[^>]*)?>").unwrap());
static RE_DIR_ATTR: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r#"([\s"'])dir="(rtl|ltr)""#).unwrap());

#[derive(Deserialize, Clone, Default)]
pub struct Translation {
    #[serde(default)]
    pub dir: String,
    #[serde(default)]
    pub values: HashMap<String, String>,
}

/// Mirrors tools/rtl-lib.mjs. Order of the two phases is load-bearing:
/// ar→X substitution first (longest-value ordering avoids a "Loading" /
/// "Loading…" prefix collision), then the lang/dir attribute patch.
pub fn substitute_and_patch(
    arabic_html: &str,
    translations: &HashMap<String, Translation>,
    from_lang: &str,
    to_lang: &str,
    to_values: &HashMap<String, String>,
    to_dir_override: &str,
) -> String {
    let empty = HashMap::new();
    let from = translations
        .get(from_lang)
        .map(|t| &t.values)
        .unwrap_or(&empty);
    // keys sorted by descending from-value length so longer strings match
    // first (longest-prefix problem); missing from-side sorts last
    let mut keys: Vec<&String> = to_values.keys().collect();
    keys.sort_by(|a, b| {
        let la = from.get(*a).map(|s| s.len()).unwrap_or(0);
        let lb = from.get(*b).map(|s| s.len()).unwrap_or(0);
        // tiebreak on the key: equal-length duplicate ar values in
        // insertion order are HashMap-order random per process, which made
        // which twin "won" (and the unmatched-warning list) nondeterministic
        lb.cmp(&la).then_with(|| a.cmp(b))
    });
    let mut unmatched: Vec<String> = Vec::new();
    let mut out = arabic_html.to_string();
    for key in keys {
        let from_val = from.get(key.as_str()).map(String::as_str).unwrap_or("");
        let to_val = to_values
            .get(key.as_str())
            .map(String::as_str)
            .unwrap_or("");
        if !from_val.is_empty() && !to_val.is_empty() && from_val != to_val {
            if !out.contains(from_val) {
                unmatched.push(key.clone());
                continue;
            }
            out = out.replace(from_val, to_val);
        } else if !to_val.is_empty() && from_val.is_empty() {
            unmatched.push(format!("{}(no {} source)", key, from_lang));
        }
    }
    if !unmatched.is_empty() {
        eprintln!(
            "warn {}: translation keys not found in HTML (left as {}): {}",
            to_lang,
            from_lang,
            unmatched.join(", ")
        );
    }
    let mut lang_dir = to_dir_override.to_string();
    if lang_dir.is_empty() {
        lang_dir = translations
            .get(to_lang)
            .map(|t| t.dir.clone())
            .unwrap_or_default();
    }
    if lang_dir.is_empty() {
        lang_dir = "ltr".to_string();
    }
    // <html lang>: replace or inject
    if RE_HTML_HAS_LANG.is_match(&out) {
        out = RE_HTML_LANG_ATTR
            .replace_all(&out, format!("${{1}}{}${{2}}", to_lang))
            .into_owned();
    } else {
        out = RE_HTML_OPEN_TAG
            .replace_all(&out, |m: &regex::Captures| {
                let m0 = m.get(0).unwrap().as_str();
                if !m0.ends_with('>') {
                    m0.to_string()
                } else {
                    format!("{} lang=\"{}\">", &m0[..m0.len() - 1], to_lang)
                }
            })
            .into_owned();
    }
    // every dir attribute (attribute-boundary anchored: a bare global would
    // also rewrite data-dir="ltr")
    out = RE_DIR_ATTR
        .replace_all(&out, format!("${{1}}dir=\"{}\"", lang_dir))
        .into_owned();
    out
}

/// The RTL demo filename convention this module writes: `-rtl-<lang>.html`
/// with lang in {"he", "en"} for every demo plus "fa" for alert. Gates and
/// sweeps classify pages through here so the set lives in one place.
pub fn rtl_page_lang(name: &str) -> Option<&'static str> {
    static R: OnceLock<Regex> = OnceLock::new();
    let re = R.get_or_init(|| Regex::new(r"-rtl-(en|he|fa)\.html$").unwrap());
    re.captures(name)
        .map(|c| match c.get(1).map(|m| m.as_str()) {
            Some("he") => "he",
            Some("fa") => "fa",
            _ => "en",
        })
}

/// The engine's own Persian dictionary (the Go source this used to live in
/// is gone); overlay's rtl:persian-dictionary audit enumerates these keys.
pub fn persian() -> HashMap<String, String> {
    [
        ("paymentTitle", "پرداخت موفق"),
        (
            "paymentDescription",
            "پرداخت حضرت به مبلغ ۲۹.۹۹ دلار با موفقیت انجام شد. رسید نیز به نشانی پست الکترونیکی شما ارسال گردید.",
        ),
        ("featureTitle", "ویژگی جدید موجود است"),
        (
            "featureDescription",
            "ما پشتیبانی از حالت تیره را به سیستم افزوده‌ایم. می‌توانید این قابلیت را از بخش تنظیمات حساب کاربری خود فعال نمایید.",
        ),
    ]
    .into_iter()
    .map(|(k, v)| (k.to_string(), v.to_string()))
    .collect()
}

pub fn run_build_rtl() -> i32 {
    let root = match std::env::current_dir() {
        Ok(r) => r,
        Err(e) => {
            eprintln!("build-rtl: {}", e);
            return 1;
        }
    };
    let dict_b = match std::fs::read_to_string(root.join("src/registry/rtl-translations.json")) {
        Ok(b) => b,
        Err(e) => {
            eprintln!("build-rtl: {}", e);
            return 1;
        }
    };
    let dict: HashMap<String, HashMap<String, Translation>> = match serde_json::from_str(&dict_b) {
        Ok(d) => d,
        Err(e) => {
            eprintln!("build-rtl: dict: {}", e);
            return 1;
        }
    };
    let tiers_b = match std::fs::read_to_string(root.join("src/registry/tiers.json")) {
        Ok(b) => b,
        Err(e) => {
            eprintln!("build-rtl: {}", e);
            return 1;
        }
    };
    #[derive(Deserialize)]
    struct TierEntry {
        #[serde(default)]
        tier: String,
        #[serde(default)]
        emit: bool,
    }
    // fail loud: an unparsable tiers.json degrading to an empty map made
    // every preview silently classify as skipped and the run exit 0
    let tiers: HashMap<String, TierEntry> = match serde_json::from_str(&tiers_b) {
        Ok(t) => t,
        Err(e) => {
            eprintln!("build-rtl: tiers.json: {}", e);
            return 1;
        }
    };
    let shipped = |name: &str| -> bool {
        match tiers.get(name.trim_end_matches("-rtl")) {
            Some(t) => t.emit || t.tier == "static" || t.tier == "kernel" || t.tier == "trivial-js",
            None => false,
        }
    };

    for d in ["docs/demos", "dist/components", "build"] {
        if let Err(e) = std::fs::create_dir_all(root.join(d)) {
            eprintln!("build-rtl: {}", e);
            return 1;
        }
    }

    let mut names: Vec<String> = dict.keys().cloned().collect();
    names.sort();

    struct Pending {
        path: std::path::PathBuf,
        html: String,
    }
    let mut pending_writes: Vec<Pending> = Vec::new();
    let mut failures: Vec<String> = Vec::new();
    let mut skipped: Vec<String> = Vec::new();
    let mut manifest: BTreeMap<String, Vec<String>> = BTreeMap::new();
    let mut emitted = 0usize;

    for name in &names {
        let existing = format!("docs/demos/{}.html", name);
        let b = match std::fs::read_to_string(root.join(&existing)) {
            Ok(b) => b,
            Err(_) => {
                if shipped(name) {
                    failures.push(format!(
                        "{}: no {} to translate from, but the component ships",
                        name, existing
                    ));
                } else {
                    skipped.push(name.clone());
                }
                continue;
            }
        };
        let empty = HashMap::new();
        let ar = dict.get(name).unwrap_or(&empty);

        let mut langs = vec!["ar".to_string()];
        let mut emit_lang = |lang: &str, values: &HashMap<String, String>, dir_override: &str| {
            let html = inject_pre_paint(&substitute_and_patch(
                &b,
                ar,
                "ar",
                lang,
                values,
                dir_override,
            ));
            for dst in [
                root.join(format!("docs/demos/{}-{}.html", name, lang)),
                root.join(format!("dist/components/{}-{}.html", name, lang)),
            ] {
                pending_writes.push(Pending {
                    path: dst,
                    html: html.clone(),
                });
            }
            emitted += 1;
            langs.push(lang.to_string());
        };
        for lang in ["he", "en"] {
            if let Some(entry) = ar.get(lang) {
                emit_lang(lang, &entry.values, "");
            }
        }
        if name == "alert-rtl" {
            emit_lang("fa", &persian(), "rtl");
        }
        manifest.insert(name.clone(), langs);
    }

    if !failures.is_empty() {
        for f in &failures {
            if let Some(i) = f.find(": ") {
                eprintln!("FAIL [{}]: {}", &f[..i], &f[i + 2..]);
            }
        }
        eprintln!(
            "FAIL  build-rtl ({} previews could not be built) — nothing written",
            failures.len()
        );
        return 1;
    }
    for p in &pending_writes {
        if let Err(e) = std::fs::write(&p.path, &p.html) {
            eprintln!("build-rtl: {}", e);
            return 1;
        }
    }
    // stale RTL pages would otherwise survive silently: nothing else sweeps
    // docs/demos (reproducible cannot see a COMMITTED file that stopped
    // regenerating, and demo-smoke excludes the RTL variants from its
    // count), while a dictionary typo turns a component into a "skipped"
    // entry. Every dictionary component regenerates here, so anything
    // -rtl- shaped that this run did not produce is stale by construction.
    let produced: std::collections::HashSet<std::path::PathBuf> =
        pending_writes.iter().map(|p| p.path.clone()).collect();
    // only this run's languages may name a deletable file: a bare `-rtl-`
    // match is wider than what build_rtl writes (-rtl-<lang>)
    let mut langs: Vec<String> = Vec::new();
    for ls in manifest.values() {
        for l in ls {
            // "ar" is the SOURCE language — no *-rtl-ar file is ever written
            if l != "ar" && !langs.contains(l) {
                langs.push(l.clone());
            }
        }
    }
    langs.sort();
    for dir in ["docs/demos", "dist/components"] {
        let Ok(ents) = std::fs::read_dir(root.join(dir)) else {
            continue;
        };
        for e in ents.flatten() {
            let p = e.path();
            let name = e.file_name().to_string_lossy().into_owned();
            let this_run_lang = langs
                .iter()
                .any(|l| name.ends_with(&format!("-{}.html", l)));
            if !name.ends_with(".html")
                || !name.contains("-rtl-")
                || !this_run_lang
                || produced.contains(&p)
            {
                continue;
            }
            if let Err(e) = std::fs::remove_file(&p) {
                eprintln!("build-rtl: removing stale {}: {}", p.display(), e);
                return 1;
            }
            println!("build-rtl: removed stale {}", p.display());
        }
    }
    let mb = match serde_json::to_string_pretty(&manifest) {
        Ok(m) => m,
        Err(e) => {
            eprintln!("build-rtl: {}", e);
            return 1;
        }
    };
    if let Err(e) = std::fs::write(root.join("build/rtl-langs.json"), mb) {
        eprintln!("build-rtl: {}", e);
        return 1;
    }
    print!(
        "build-rtl: {} language variants emitted (excluding ar default) + manifest for {} previews",
        emitted,
        manifest.len()
    );
    if !skipped.is_empty() {
        print!(
            " ({} dictionaries have no shipped page: {})",
            skipped.len(),
            skipped.join(", ")
        );
    }
    println!();
    0
}
